### PR TITLE
JS: recognize more HttpResponseSink

### PR DESCRIPTION
**Pull request by erik-krogh**
In this example: 
```JavaScript
app.get('/user/:id', function (req, res) {
  if (whatever) {
    res.set('Content-Type', 'text/plain');
    res.send("FOO: " + req.params.id); // OK - content type is plain text
  } else {
    res.set('Content-Type', 'text/html');
    res.send("FOO: " + req.params.id); // NOT OK - content type is HTML. Previously OK
  }
});
```

We would previously not recognize any of the `res.send` as a sink, as there existed a non-html `Content-Type` for the route-handler. 

This PR tries to recognize if a HeaderDefinition (e.g. `res.set(..)`) can affect a ResponseSendArgument (e.g. `res.send(..)`). 

We gain [plenty of new sinks with this change](https://lgtm.com/query/9220294355482981469/). (All of which are real sinks, except for 1 which was fixed by the drive-by-change mentioned below). 

This change gets us a TP in CVE-2018-16480, CVE-2018-3747, and CVE-2018-16484, and recognizes the sink in CVE-2018-3726. 

The drive-by-change to `StandardRouteHandler::getAResponseHeader` allows getAResponseHeader to return headers that does not extend `StandardHeaderDefinition`. 
For example: `SetMultipleHeaders` from `Express.qll`. 

[Evaluation has been mostly good](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1582925430664), but with weird performance (ignore `azure-sdk-for-node`, that looked like a cached result). 
But I couldn't replicate the other performance regressions by running codeql locally.   
I'm doing security-suite on nightly, and I'll see how that works out.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll [hunk 1]
```diff
@@ -305,18 +305,64 @@ module ReflectedXss {
...
-      not exists(string tp | hd.defines("content-type", tp) | tp.regexpMatch("(?i).*html.*"))
+  HTTP::HeaderDefinition nonHtmlContentTypeHeader(HTTP::RouteHandler h) {
+    result = h.getAResponseHeader("content-type") and
+    not exists(string tp | result.defines("content-type", tp) | tp.regexpMatch("(?i).*html.*"))
+  }
+
+  /**
+   * Holds if a header set in `header` is unlikely to affect a resonse send in `sender`.
+   */
+  predicate isIrrelevantFor(HTTP::HeaderDefinition header, HTTP::ResponseSendArgument sender) {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll]: I'm not comfortable with this predicate.

In terms of  performance, it appears to be a cartesian product of all header definitions and sender arguments with some tuples subtracted. Indeed, the second disjunct below doesn't bind `sender`.

I'm also a little worried about the second disjunct being too dependent on everything being in a final basic block containing no resolved calls.

Have you tried using dominance for the local control-flow checks? I think we could define it all in terms of dominance relations, like:
- A send is affected by its dominating header definition.
- A dominating header definition is "local" if it dominates the response `end` call.
- A send without a dominating header definition is assumed to be affected by any non-local header definition on the same response object (i.e. a flow-insensitive heuristic for the non-local case).

Lastly, content-type tracking seems like an important enough concept to justify support in the `HTTP` library, so if you feel like building a more detailed analysis (akin to the middleware-tracking in Express) I'd be in favor of that.

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll]: > Have you tried using dominance for the local control-flow checks? I think we could define it all in terms of dominance relations, like:
> 
> * A send is affected by its dominating header definition.
> * A dominating header definition is "local" if it dominates the response `end` call.
> * A send without a dominating header definition is assumed to be affected by any non-local header definition on the same response object (i.e. a flow-insensitive heuristic for the non-local case).

I tried out an implementation based on your idea, and it works nicely :+1: 

> Lastly, content-type tracking seems like an important enough concept to justify support in the `HTTP` library, so if you feel like building a more detailed analysis (akin to the middleware-tracking in Express) I'd be in favor of that.

I don't see the need for such a detailed analysis yet. `Content-type` tracking is currently only used for this `XSS` sink, so I want to wait with such an analysis until we have another use for `Content-type`.

**Comment by erik-krogh** [comment]
[The new sinks](https://lgtm.com/query/352372578652716087/) still look good after the changes.

**Comment by erik-krogh** [comment]
A [new evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1583282427105) after the changes still look good.  
The new result is a TP. 

I'll see if I can find an explanation for the performance regression on Rhino.

**Comment by erik-krogh** [comment]
> I'll see if I can find an explanation for the performance regression on Rhino.

The performance regression was simply caused by the query using `BasicBlock::dominates`, which `XSS` queries didn't use previously.   
Adding another query that already used `BasicBlock::dominates` into the evaluation causes the performance regression to disappear.

**Pull request merged**